app/models/concerns/updatable.rb

```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module Updatable
+  extend ActiveSupport::Concern
+
+  def updatable_parent_object(oid, index)
+    parent_object = ParentObject.find_by(oid: oid)
+    if parent_object.blank?
+      batch_processing_event("Skipping row [#{index + 2}] with parent oid: #{oid} because it was not found in local database", 'Skipped Row')
+      return false
+    elsif !current_ability.can?(:update, parent_object)
+      batch_processing_event("Skipping row [#{index + 2}] with parent oid: #{oid}, user does not have permission to update.", 'Permission Denied')
+      return false
+    else
+      parent_object
+    end
+  end
+
+  def update_parent_objects
+    return unless batch_action == "update parent objects"
+    parsed_csv.each_with_index do |row, index|
+      oid = row['oid'] unless ['oid'].nil?
+      metadata_source = row['source'] unless row['source'].nil?
+      parent_object = updatable_parent_object(oid, index)
+      next unless parent_object
+
+      processed_fields = process_fields_for_update(parent_object, row)
+      setup_for_background_jobs(parent_object, metadata_source)
+      parent_object.update(processed_fields)
+
+      processing_event_for_parent(parent_object)
+    end
+  end
+
+  # rubocop:disable Style/MultilineTernaryOperator
+  # rubocop:disable Style/TernaryParentheses
+  def process_fields_for_update(parent_object, row)
+    fields = ['aspace_uri', 'barcode', 'bib', 'digitization_note', 'holding', 'item', 'rights_statement']
+    validation_fields = { "display_layout" => 'viewing_hints', "extent_of_digitization" => 'extent_of_digitizations', "viewing_direction" => 'viewing_directions', "visibility" => 'visibilities' }
+
+    processed_fields = {}
+    fields.each do |f|
+      processed_fields[f.to_sym] = (
+        row[f].present? &&
+        row[f] != parent_object.send(f)
+      ) ? row[f] : parent_object.send(f)
+    end
+    validation_fields.each do |k, v|
+      processed_fields[k.to_sym] = (
+        row[k].present? &&
+        row[k] != parent_object.send(k) &&
+        (ParentObject.send(v).include? row[k])
```

Comment:
You should probably add a batch message if the validation fails for one of the fields.
I wonder also if you should reject the whole row if one of the fields fails validation instead of skipping just that column.

Reply:
We definitely need to let the user know so they can correct the problem and resubmit for that OID.

Reply:
![Image 2021-08-12 at 4 28 04 PM](https://user-images.githubusercontent.com/24666568/129282720-c72f873e-6790-4ad5-96d5-8e1c2c55f04b.jpg)

Reply:
We've added some detailed batch messages.  And are just cleaning up the specs now.

![image](https://user-images.githubusercontent.com/36549923/129284221-d6c5b64b-b450-4467-a392-fa0f07da1aa8.png)

Reply:
Nice messages!